TELCODOCS-1576: Image-based upgrade with Lifecycle Agent (#69809)
openshift/openshift-docs: modules/ztp-image-based-upgrade-generate-seed-image.adoc

* You are logged in as a user with `cluster-admin` privileges.
* The seed cluster has the same CPU topology as the target cluster.
* The seed cluster is registered as a managed cluster.
* The {lcao} deployed on the target cluster is compatible with the version in the seed image.
// Are there more prereqs?

.Procedure

. Detach the seed cluster from the hub cluster either manually or if using ZTP, by removing the `SiteConfig` CR from the `kustomization.yaml`.

.. To manually detach the seed cluster, run the following command:
+
[source,terminal]
----
$ oc delete managedcluster sno-worker-example
----

.. Wait until the `ManagedCluster` CR is removed. Once the CR is removed, create the proper `SeedGenerator` CR. The {lcao} cleans up the {rh-rhacm} artifacts.

.. If you are using GitOps ZTP, remove the seed cluster's `SiteConfig` CR from the `kustomization.yaml`:
+
[source,yaml]
----
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

generators:
#- example-seed-sno1.yaml
- example-target-sno2.yaml
- example-target-sno3.yaml
----

.. Commit the `kustomization.yaml` changes in your Git repository and push the changes.
+

> The ArgoCD pipeline detects the changes and removes the managed cluster.

Reviewer: Thanks, this makes much more sense

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: PROCEDURE
[id="ztp-image-based-upgrade-seed-generation_{context}"]
= Generating a seed image with the {lcao}

Use the {lcao} to generate the seed image with the `SeedGenerator` CR. The Operator does the following tasks:

// seed cluster, some things are cluster specific that are not included
// all operator, LCAO, OADP, LVMS, included in the seed image, telco day2 stuff is incl
// system config checks ()

* Checks for required system configurations
* Performs any necessary system cleanup before generating the seed image
* Launches the image generation, which involves the following tasks:
** Stops cluster operators
** Prepares the seed image configuration

> ** Generates and pushes the seed image to the image repository specified in the `ImageBasedUpgrade` CR

Reviewer 2: specified in the SeedGenerator CR

** Restores cluster operators
** Expires seed cluster certificates

> * Generates new certificates for the seed cluster

Reviewer 2: Shouldn't this and the next item also have 2 asterisks?

* Restores and updates the `SeedGenerator` CR on the seed cluster, and the `ManagedCluster` CR on the hub cluster

[IMPORTANT]
====
The generated seed image does not include any site-specific data.
====

//TODO Expand on what the seed image contains and explicitly say what it doesn't

.Prerequisites

* Deploy a {sno} cluster with a DU profile.
* Install the {lcao} on the seed cluster.

> * Install the OADP Operator on the seed cluster.

Reviewer 2: Since it's optional to install OADP in the target cluster - maybe should be optional here too.

Author: Clarified since then, that it's not optional. I'm currently cleaning up the docs now regarding this

* Log in as a user with `cluster-admin` privileges.
* The seed cluster has the same CPU topology as the target cluster.
* If using {rh-rhacm}, the seed cluster is registered as a managed cluster.
* The {lcao} deployed on the target cluster is compatible with the version in the seed image.
// Are there more prereqs?

.Procedure

> . (Optional) If you use {rh-rhacm} to manage your clusters, detach the seed cluster from the hub cluster either manually or if using ZTP, by removing the `SiteConfig` CR from the `kustomization.yaml`.

Reviewer 2: @alosadagrande @browsell
Is this still required?

Author: Brent told me that hubKubeconfig should not be in the customer docs, these two options are needed instead

----

[source,terminal]
----
$ base64 -w 0 ${AUTHFILE} ; echo
----
--

.. Copy the output into the `seedAuth` field in the `Secret` CR.

. Apply the `Secret` CR.
+
[source,terminal]
----
$ oc apply -f secretseedgenerator.yaml
----

. Create the `SeedGenerator` CR:
+
--
[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: SeedGenerator
metadata:
  name: seedimage <1>
spec:
  seedImage: <seed_container_image> <2>
----
<1> The `SeedGenerator` CR must be named `seedimage`.
<2> Specify the container image URL, for example, `quay.io/example/seed-container-image:<tag>`. It is recommended to use the `<seed_cluster_name>:<ocp_version>` format.
--

. Generate the seed image by running the following command:
+
[source,terminal]
----
$ oc apply -f seedgenerator.yaml
----

+
[IMPORTANT]
====
The cluster reboots and loses API capabilities while the {lcao} generates the seed image.
Applying the `SeedGenerator` CR stops the `kubelet` and the CRI-O operations, then it starts the image generation.
====

Once the image generation is complete, the cluster can be reattached to the hub cluster, if applicable, and you can access it through the API.

> If you want to generate further seed images, you must provision a new seed cluster with the version you want to generate a seed image from.

Reviewer 2: AFAIK, we support generating seed  images from the same seed cluster too.

.Verification

. Once the cluster recovers and it is available, you can check the status of the `SeedGenerator` CR:
+
--
[source,terminal]
----
$ oc get seedgenerator -A -oyaml
----

.Example output
[source,yaml]
----
status:

>     conditions:

Reviewer 2: In this example the time of type: SeedGenInProgress is later then type: SeedGenCompleted.

here's an example from my setup:
`
status:
  conditions:
  - lastTransitionTime: "2024-02-13T21:24:26Z"
    message: Seed Generation completed
    observedGeneration: 1
    reason: Completed
    status: "False"
    type: SeedGenInProgress
  - lastTransitionTime: "2024-02-13T21:24:26Z"
    message: Seed Generation completed
    observedGeneration: 1
    reason: Completed
    status: "True"
    type: SeedGenCompleted
  observedGeneration: 1

`

----

. Create the `SeedGenerator` CR:
+
--
[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: SeedGenerator
metadata:
  name: seedimage <1>
spec:
  seedImage: <seed_container_image> <2>
----
<1> The `SeedGenerator` CR must be named `seedimage`.
<2> Specify the container image URL, for example, `quay.io/example/seed-container-image:<tag>`. It is recommended to use the `<seed_cluster_name>:<ocp_version>` format.
--

. Generate the seed image by running the following command:
+
[source,terminal]
----
$ oc apply -f seedgenerator.yaml
----

+
[IMPORTANT]
====
The cluster reboots and loses API capabilities while the {lcao} generates the seed image.
Applying the `SeedGenerator` CR stops the `kubelet` and the CRI-O operations, then it starts the image generation.
====

Once the image generation is complete, the cluster can be reattached to the hub cluster, and you can access it through the API.

If you want to generate further seed images, you must provision a new seed cluster with the version you want to generate a seed image from.

.Verification

. Once the cluster recovers and it is available, you can check the status of the `SeedGenerator` CR:
+
--
[source,terminal]
----

> $ oc get seedgenerator -A -oyaml

Reviewer 3: ```suggestion
$ oc get seedgenerator -oyaml
```

```
-> oc api-resources --namespaced=false | grep seedgenerator
seedgenerators                    seedgen             lca.openshift.io/v1alpha1                     false        SeedGenerator
```

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: PROCEDURE
[id="ztp-image-based-upgrade-seed-generation_{context}"]
= Generating a seed image with the {lcao}

Use the {lcao} to generate the seed image with the `SeedGenerator` CR. The {lcao} will do the following:

> // seed cluster, some things are cluster specific that are not included

Reviewer 4: What about any 3rd party parts like drivers that are installed with KMM?

Reviewer 5: out of scope

Reviewer 4: Half of production deployments use 3rd party drivers or libraries. This is not the answer we can tell them. You are giving them excuse to dump RH and go with a different vendor.

--

.. Create the authentication and the `Kubeconfig` files by running the following commands:
+
--
.Authentication file
[source,terminal]
----
$ MY_USER=myuserid
$ AUTHFILE=/tmp/my-auth.json
$ podman login --authfile ${AUTHFILE} -u ${MY_USER} quay.io/${MY_USER}
----

[source,terminal]
----
$ base64 -w 0 ${AUTHFILE} ; echo
----

.hubKubeconfig file
[source,terminal]
----
$ HUBKUBECONFIGFILE=/home/<user_ID>/hubkubeconfig
----

[source,terminal]
----
$ base64 -w 0 ${HUBKUBECONFIGFILE} ; echo
----
--

.. Copy the outputs into the respective `seedAuth` and `hubKubeconfig` fields in the `Secret` CR.

. Apply the `Secret` CR.
+
[source,terminal]
----
$ oc apply -f secretseedgenerator.yaml
----

. (Optional) If you did not populate the `hubKubeconfig` field in the `Secret` CR, detach the seed cluster from the hub cluster and remove all {rh-rhacm} and multicluster engine certificates that cannot be recertificated in the target cluster.

.. Detach the seed cluster from the hub cluster:
+
[source,terminal]
----
$ oc delete managedcluster sno-worker-example
----

> .. Wait until the `ManagedCluster` CR is removed. Once the CR is removed, create the proper `SeedGenerator` CR. The {lcao} cleans up the {rh-rhacm} artifacts.

Reviewer 4: somewhat outside of this update but if customer want to resize their cluster do we indicate when we recommend them to do it? before LAO starts, after update completed or somewhere here when managecluster is deleted?

Reviewer 5: Can you elaborate on resize ?  This is a SNO.

Reviewer 4: If it is just for SNO then I will drop my comment.

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: PROCEDURE
[id="ztp-image-based-upgrade-seed-generation_{context}"]
= Generating a seed image with the {lcao}

Use the {lcao} to generate the seed image with the `SeedGenerator` CR. The Operator does the following:

// seed cluster, some things are cluster specific that are not included

> // all operator, LCAO, OADP, LVMS, included in the seed image, telco day2 stuff is incl

Reviewer 5: We will need to expand on this

Author: @browsell Could you please point me in the right direction on what would you like to see here exactly?

// system config checks ()

* Checks for required system configurations
* Performs any necessary system cleanup before generating the seed image
* Launches the image generation task, which involves the following:
** Stops cluster operators
** Prepares the seed image configuration
** Generates and pushes the seed image to the image repository specified in the `ImageBasedUpgrade` CR
** Restores cluster operators
** Expires seed cluster certificates
* Generates new certificates for the seed cluster
* Restores and updates the `SeedGenerator` CR on the seed cluster, and the `ManagedCluster` CR on the hub cluster

//TODO Expand on what the seed image contains and explicitly say what it doesn't

.Prerequisites

> * You deployed a {sno} with DU profile.

Reviewer 6: {sno} cluster

Reviewer 5: a partial DU profile, any site specific data cannot be included

* The seed cluster is registered as a managed cluster.
* The {lcao} deployed on the target cluster is compatible with the version in the seed image.
// Are there more prereqs?

.Procedure

. Detach the seed cluster from the hub cluster either manually or if using ZTP, by removing the `SiteConfig` CR from the `kustomization.yaml`.

.. To manually detach the seed cluster, run the following command:
+
[source,terminal]
----
$ oc delete managedcluster sno-worker-example
----

.. Wait until the `ManagedCluster` CR is removed. Once the CR is removed, create the proper `SeedGenerator` CR. The {lcao} cleans up the {rh-rhacm} artifacts.

.. If you are using GitOps ZTP, remove the seed cluster's `SiteConfig` CR from the `kustomization.yaml`:
+
[source,yaml]
----
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

generators:
#- example-seed-sno.yaml
- example-target-sno1.yaml
- example-target-sno1.yaml
----

.. Commit the `kustomization.yaml` changes in your Git repository and push the changes.
+
The ArgoCD pipeline detects the changes and removes the managed cluster.

. Create the `Secret` CR in the `openshift-lifecycle-agent` namespace.

> +

Reviewer 6: I found the step alignment a bit confusing here. Should it be:
1. Create the secret 
 a. Step 1 create a yaml file etc.
 b. create auth
 c.  Copy the secret.
 d. apply the secret CR
 2. Next task..

+
--
.hubKubeconfig file
[source,terminal]
----
$ HUBKUBECONFIGFILE=/home/<user_ID>/hubkubeconfig
----

[source,terminal]
----
$ base64 -w 0 ${HUBKUBECONFIGFILE} ; echo
----
--

.. Copy the outputs into the respective `seedAuth` and `hubKubeconfig` fields in the `Secret` CR.

. Apply the `Secret` CR.
+
[source,terminal]
----
$ oc apply -f secretseedgenerator.yaml
----

. Create the `SeedGenerator` CR:
+
--
[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: SeedGenerator
metadata:
  name: seedimage <1>
spec:
  seedImage: <seed_container_image> <2>
----
<1> The `SeedGenerator` CR must be named `seedimage`.
<2> Specify the container image URL, for example, `quay.io/example/seed-container-image:<tag>`. It is recommended to use the `<seed_cluster_name>:<ocp_version>` format.
--

. Generate the seed image by running the following command:
+
[source,terminal]
----
$ oc apply -f seedgenerator.yaml
----

+
[IMPORTANT]
====
The cluster reboots and loses API capabilities while the {lcao} generates the seed image.
Applying the `SeedGenerator` CR stops the `kubelet` and the CRI-O operations, then it starts the image generation.
====

> Once the image generation is complete, the cluster is reattached to the hub cluster and can be accessed through the API.

Reviewer 7: does this happen automatically even if someone didn't supply the optional hub kubeconfig, or would they have to manually add it back in to the cluster?

Author: It does not happen automatically, I just revised this part to make it obvious it's a manual reattachment.

I also clarified with the team that the hub kubeconfig option is not customer facing, so any mention of that has been removed, ([see here](https://69809--ocpdocs-pr.netlify.app/openshift-enterprise/latest/scalability_and_performance/ztp_far_edge/ztp-talm-updating-managed-policies#ztp-image-based-upgrade-seed-generation_ztp-talm))

// Module included in the following assemblies:

> // Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576

Reviewer 8: nit - these comments aren't required. We can see all this from the git history.

// seed cluster, some things are cluster specific that are not included
// all operator, LCAO, OADP, LVMS, included in the seed image, telco day2 stuff is incl
// system config checks ()

* Checks for required system configurations
* Performs any necessary system cleanup before generating the seed image
* Launches the image generation task, which involves the following:
** Stops cluster operators
** Prepares the seed image configuration
** Generates and pushes the seed image to the image repository specified in the `ImageBasedUpgrade` CR
** Restores cluster operators
** Expires seed cluster certificates
* Generates new certificates for the seed cluster
* Restores and updates the `SeedGenerator` CR on the seed cluster, and the `ManagedCluster` CR on the hub cluster

//TODO Expand on what the seed image contains and explicitly say what it doesn't

.Prerequisites

* You deployed a {sno} with DU profile.
* You installed the {lcao} on the seed cluster.
* You installed the OADP Operator on the seed cluster.
* You are logged in as a user with `cluster-admin` privileges.
* The seed cluster has the same CPU topology as the target cluster.

> * The seed cluster and the target cluster have the same {rh-rhacm} or multicluster engine version.

Reviewer 5: This is not a prereq

//seedgen Secret and seedgenerator get deleted before imager to not be added to the image
** Stops the cluster operators
** Prepares the seed image configuration
** Generates and pushes the seed image to the image repository specified in the `imageBasedUpgrade` CR
** Restores the cluster operators
** Expires seed cluster certifications
//** Prepares for precaching operations
//** Backs up platform configurations and artifacts
* Generates new certificates for the seed cluster
* Restores and updates the `SeedGenerator` CR on the seed cluster, and the `ManagedCluster` CR on the hub cluster

//TODO Expand on what the seed image contains and explicitly say what it doesn't

.Prerequisites

* You deployed an {sno} with DU profile.
* You installed the {lcao} on the seed cluster.
* You installed the OADP Operator on the seed cluster.
* You are logged in as a user with `cluster-admin` privileges.
* The seed cluster has the same CPU topology as the target cluster.
* The seed cluster and the target cluster have the same {rh-rhacm} or multicluster engine version.
* The seed cluster is registered as a managed cluster.
// Are there more prereqs?

.Procedure

> . Detach the seed cluster.

Reviewer 9: Detach the seed cluster either manually or by providing the hubKubeConfig information as shown in the next step. It should be somehow linked with the next step, otherwise, a user could detach the cluster and then add the hubkubeconfig in the next step too.

Author: I expanded this to be super clear and now the two ways to detach the cluster is mentioned in two places:

> Detach the seed cluster manually or provide the `hubKubeConfig` information in the `Secret` CR.

The next step is creating the CR and I also highlight in callout 3 that you need to do it manually or populate the field.

  name: seedimage <1>
spec:
  seedImage: <seed_container_image> <2>
  recertImage: "quay.io/edge-infrastructure/recert:latest"
----
<1> The `SeedGenerator` CR must be named `seedimage`.
<2> Specify the container image URL, for example, `quay.io/example/seed-container-image:<tag>`. It is recommended to use the `<seed_cluster_name>:<ocp_version>` format.

[IMPORTANT]
====
The same version of the applications must function on both the current and the target release of {product-title}.
====
--

. Generate the seed image by running the following command:
+
[source,terminal]
----
$ oc apply -f seedgenerator.yaml
----

+
[IMPORTANT]
====
The cluster reboots and loses API capabilities while the {lcao} generates the seed image.
====

+
Applying the `SeedGenerator` CR stops the `kubelet` and the CRI-O operations, and then it starts the image generation.
Finally, the {lcao} starts a cleanup process that allows you to keep the seed {sno} fully functional.
//This might not be important info, check

.Verification

. You can monitor the image generation process.
// this sounds all kind of wrong, check where you can monitor it exactly

.. SSH into the seed cluster:
+
[source,terminal]
----

> $ ssh <node_ID>

Reviewer 9: ssh core@<node_ID>

Not sure if this is more troubleshooting than verification. I mean, verification must be led by the status of the seedgenerator CR. Since the cluster is not available during the image creation process, the user should wait for the cluster to be up and running again. In case the cluster is not responsive after some time, we could check if there was any issue during the image creation process by connecting as shown in the example.  Wdyt?

Author: I chatted with someone about creating a dedicated Troubleshooting section, I think that would be cleaner. Once we established some common scenarios, I'm happy to create that section, so I can put this there and expand depending on the issue. For now, I'm going to comment it out and add a TODO for myself. Keeping the CR monitoring step in Verification

. Apply the `Secret` CR.
+
[source,terminal]
----
$ oc apply -f secretseedgenerator.yaml
----

. (Optional) If you did not populate the `hubKubeconfig` field in the `Secret` CR, detach the seed cluster from the hub cluster and remove all {rh-rhacm} and multicluster engine certificates that cannot be recertificated in the target cluster.

.. Detach the seed cluster from the hub cluster:
+
[source,terminal]
----
$ oc delete managedcluster sno-worker-00
----

.. Wait until the `ManagedCluster` CR is removed. Once the CR is removed and you create the proper `SeedGenerator` CR, the {lcao} cleans up the {rh-rhacm} artifacts.

. Create the `SeedGenerator` CR:
+
--
[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: SeedGenerator
metadata:
  name: seedimage <1>
spec:
  seedImage: <seed_container_image> <2>
#  recertImage: <recertification_tool_container_image> <3>
----
<1> The `SeedGenerator` CR must be named `seedimage`.
<2> Specify the container image URL, for example, `quay.io/example/seed-container-image`.
//<3> Optional. Specify the recertification tool container image to pass to the {lcao}, for example: `quay.io/example/recert`.

[IMPORTANT]
====

> The same version of the applications must function on both the current and the target release of {product-title}.

Author: Where would it fit better? I'm making an assumption here, but wouldn't the user need to check this before generating the seed image? "Need" meaning that they're recommended to do this before spending time generating the image

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: PROCEDURE
[id="ztp-image-based-upgrade-seed-generation_{context}"]
= Generating a seed image with the {lcao}

Use the {lcao} to generate the seed image with the `SeedGenerator` CR. The {lcao} will do the following:

* Checks for required system configurations
* Performs any necessary system cleanup before generating the seed image

> * Launches the image generation tool, which does the following:

Reviewer 10: >I think it is important to notice two things here as well:

Did you mean to say "note" instead of "notice"?

> * Launches the image generation tool, which does the following:

Reviewer 9: correct, thx

> * Launches the image generation tool, which does the following:

Reviewer 10: >I think it is important to note two things here as well:
The image generation tool or task calls recert (not sure if it is going to be productized) which is in charge of expiring all the crypto information stored in the seed cluster. cc/ @omertuc
That's important so private crypto information of a running/valid SNO (the seed cluster) cannot be obtained by pulling the seed container image. It can be the case that could obtain the kubeconfig /credentials to log into the seed cluster.

That's a bit of an implementation detail, I don't want users to know what recert is, or even think about their crypto, so I'm against including that in the doc text

> * Launches the image generation tool, which does the following:

Reviewer 9: I agree, the idea is to give @amolnar-rh more detail. However, I think it is important to let the users know that the cryptographic certificates are expired in the seed and regenerated again in the target. The users do not have to worry about the container seed image being obtained by a non-trusted person and then it can extract or obtain credentials/keys of the seed and connect or add some malicious code.

> * Launches the image generation tool, which does the following:

Reviewer 10: >The users do not have to worry

Sure, if we want to communicate that assurance then we could have something like that then.

Just one nit: 

> the cryptographic certificates are expired

Technically expiring the certificates is not the reason the user shouldn't have to worry, the thing that recert does that makes it safe is replacing all the private keys / JWTs with different random ones. Also let's say that this happens but without mentioning recert, just that it happens

> * Launches the image generation tool, which does the following:

Author: Okay, so it's clear that the recert tool shouldn't be mentioned, but it's a good idea to say something about the recertification process happening. I currently have an overview of the process, which mentions it as:
> The LCA will do the following:
> [...]
> * Launches the image generation task, which involves the following:
> ** Stops the cluster operators
> ** Prepares the seed image configuration
> ** Generates and pushes the seed image to the image repository specified in the `imageBasedUpgrade` CR
> ** Restores the cluster operators
> ** Expires seed cluster certifications
> * Generates new certificates for the seed cluster
> [...]

I have a few references to the recert tool, I'll remove them in the next commit.

> ** Stops the cluster operators

Reviewer 9: what does it mean stops clusters operators? IIRC it stops the kubernetes agent (kubelet) and the container runtime (crio) so, then all Pods running on OpenShift are stopped too. I mean, creates a clean environment to create the image. cc/ @leo8a

** Prepares the seed image configuration

> ** Generates and pushes the seed image to the image repository

Reviewer 9: ...to the image repository specified in the `imageBasedUpgrade` custom resource or in the repository specified by the user. IMHO I think that way we can link the information explained at the beginning of the section where you explain that there are two main CRs.

> ** Restores the cluster operators

Reviewer 9: restore the cluster?

//** Prepares for precaching operations
//** Backs up platform configurations and artifacts
* Restores and updates the `SeedGenerator` CR, and restores the `ManagedCluster` CR on the hub cluster

> 

Reviewer 9: Probably add a note to warn users that the cluster will be up and running. But you are not suggested to upgrade using a regular way (oc adm upgrade) and create a new seed image from the new version and so on. cc/ @browsell @omertuc 

AFAIK the suggested way is to install and configure a new seed cluster from scratch. I would not include that sentence, but I will include that it is not suggested (supported?) to run multiple seed image creations of the same seed cluster,e .g: run seed image creation, upgrade seed cluster, run seed image creation, upgrade cluster again...

* The seed cluster has the same CPU topology as the target cluster.
* The seed cluster and the target cluster have the same {rh-rhacm} or multicluster engine version.
* The seed cluster is registered as a managed cluster.
// Are there more prereqs?

.Procedure

. Apply the shared container folder `MachineConfig` CR to make `/var/lib/containers` shared between `ostree` updates.
+
[source,terminal]
----
$ oc apply -f <shared_MC_folder>
----

// TODO Remove becaue dnsmasq is temprorary
. Generate the proper `dnsmasq` configuration and save the output to a file.
+
[source,terminal]
----
$ ./generate-dnsmasq-site-policy-section.sh --name <seed_name> --domain <seed_domain> --ip <seed_IP> --mc > mc-sno1-dnsmasq-seed.yaml
----

. Apply the `dnsmasq` configuration.
+
[source,terminal]
----
$ oc apply -f mc-sno1-dnsmasq-seed.yaml
----

+
The seed cluster reboots.

. Create the `Secret` CR in the `openshift-lifecycle-agent` namespace.
+
--
[source,yaml]
----
apiVersion: v1
kind: Secret
metadata: <1>
  name: seedgen
  namespace: openshift-lifecycle-agent
type: Opaque
data:
  seedAuth: <encoded_AUTHFILE> <2>
  hubKubeconfig: <encoded_KUBECONFIG> <3>
----

> <1> The `Secret` CR must have the `name: seedgen` and `namespace: openshift-lifecycle-agent` fields.

Author: I explicitly state this in the docs here as callout <3> and an optional step, step 4, details to detach the cluster and why you have to do it.

. Create the `SeedGenerator` CR:
+
--
[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: SeedGenerator
metadata:
  name: seedimage <1>
spec:
  seedImage: <seed_container_image> <2>
#  recertImage: <recertification_tool_container_image> <3>
----
<1> The `SeedGenerator` CR must be named `seedimage`.
<2> Specify the container image URL, for example, `quay.io/example/seed-container-image`.
//<3> Optional. Specify the recertification tool container image to pass to the {lcao}, for example: `quay.io/example/recert`.

[IMPORTANT]
====
The same version of the applications must function on both the current and the target release of {product-title}.
====
--

. Generate the seed image by running the following command:
+
[source,terminal]
----
$ oc apply -f seedgenerator.yaml
----

+
[IMPORTANT]
====
The cluster loses API capabilities while the {lcao} generates the seed image.
====

+
Applying the `SeedGenerator` CR stops the `kubelet` and the CRI-O operations, and then it starts the image generation.
Finally, the {lcao} starts a cleanup process that allows you to keep the seed {sno} fully functional.

> 

Reviewer 9: As stated in a previous comment, yes, the seed cluster is fully functional again. But it is not supported to upgrade the seed cluster and run another seed image creation. User must provision a new seed cluster with the proper version they want to create the seed image.

.Verification

. You can monitor the image generation process from the target cluster.

.. SSH into the target cluster:
+
[source,terminal]
----
$ ssh <node_ID>
----
// must have an ssh key configured via MC. When do we do this and how?

.. List the running containers:
+
--
[source,terminal]
----
$ podman ps
----

.Example output
[source,terminal]
----
CONTAINER ID  IMAGE                                             COMMAND               CREATED         STATUS         PORTS       NAMES
f9484b2e64dc  quay.io/<user>/lifecycle-agent-operator:latest  create --authfile...  31 seconds ago  Up 31 seconds              ibu_imager
----
--

.. Check the image generation logs by running the following command:
+
[source,terminal]
----
$ podman logs -f ibu_imager
----

> . Check the status of the `SeedGenerator` CR:

Reviewer 9: Once the cluster is back online, check the status of the....

    - lastTransitionTime: "2024-01-01T09:00:00Z"
      message: Seed Generation completed
      observedGeneration: 1
      reason: Completed
      status: "True"
      type: SeedGenCompleted <1>
    - lastTransitionTime: "2024-01-01T09:02:00Z"
      message: Seed Generation completed
      observedGeneration: 1
      reason: Completed
      status: "False"
      type: SeedGenInProgress
    observedGeneration: 1
----
<1> The seed image generation is complete.
--

. Verify that the {sno} is running and is attached to the {rh-rhacm} hub cluster:
+
--
[source,terminal]
----
$ oc get nodes,clusterversion,co
----

.Example output
[source,terminal]
----
NAME                                            STATUS   ROLES                         AGE   VERSION
node/snonode.sno-worker-02.example.redhat.com   Ready    control-plane,master,worker   12m   v1.27.6+d548052

NAME                                         VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS
clusterversion.config.openshift.io/version   4.14.4    True        False         11h     Cluster version is 4.14.4
----

[NOTE]
====
The {sno} node requires time to recover after restarting the `kubelet` operation.
====
--

> [id="ztp-image-based-upgrade-additional-seed-image-generation_{context}"]

Reviewer 9: Is this true? cc/ @browsell  @omertuc  - I understood that it is not going to be supported.

Reviewer 10: True, section can be removed. Seed clusters should be considered spoiled after they live for more than a day or so. Right now it's in theory, in the future it'll be a hard error once we make a final decision about it.

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: PROCEDURE
[id="ztp-image-based-upgrade-seed-generation_{context}"]
= Generating a seed image with the {lcao}

Use the {lcao} to generate the seed image. The image generation does the following:

* Prepares for pre-caching operations

> * Stops the `kubelet` service, running containers, and the CRI-O runtime

Reviewer 8: ```suggestion
* Stops the `kubelet` service, running containers, and the CRI-O runtime on the managed cluster
```

* Backs up platform configurations and artifacts
* Creates and pushes the OCI images

[IMPORTANT]
====
The same version of the applications must function on both the current and the target release of {product-title}.
====

.Prerequisites

> * You installed the {lcao}.

Reviewer 8: ```suggestion
* You installed the {lcao} on the hub cluster.
```

> * You have an existing {sno} cluster.

Reviewer 8: ```suggestion
* The cluster you want to upgrade is registered as a managed cluster.
```

> // Are there more prereqs?

Reviewer 8: Log in as administrator on the hub cluster?

.Procedure

. Create and edit the `SeedGenerator` CR:
+
[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: SeedGenerator
metadata:
  name: seedimage
spec:
  seedImage: quay.io/<user>/<seed_image> <1>

>   recertImage: quay.io/<repository>/recert:latest <2>

Reviewer 8: Also you don't have any other info about this recertification tool. What is it? what does it do? Do you have to use it? etc.

----

> <1> Specifies the repository that is used to pull the generated image from.

Reviewer 8: General comment: You don't have much info about what this seed image entails, what it contains, can you run it locally, is there a sample containerfile, etc.

<2> Optional. Allows you to specify the recertification tool image to pass to the {lcao}.

. Create and edit the `Secret` CR in the `openshift-lifecycle-agent` namespace:
+
[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: seedgen
  namespace: openshift-lifecycle-agent
type: Opaque
data:
  seedAuth: <encoded_AUTHFILE> <1>
  hubKubeconfig: <encoded_KUBECONFIG> <2>
----
<1> Specifies a `base64`-encoded authfile for write-access to the registry for pushing the generated seed images.

> <2> Optional. Specifies a `base64`-encoded `KUBECONFIG` file for admin access to the hub to deregister the seed cluster from {rh-rhacm}. If this is not present in the secret, the {rh-rhacm} cleanup is skipped.

Reviewer 8: ```suggestion
<2> Optional. Specifies a base64-encoded `KUBECONFIG` file for administrator access to the hub cluster to unregister the seed cluster from {rh-rhacm}. If this is not present in the secret, the {rh-rhacm} cleanup is skipped.
```

. To generate the seed image, run the following command:
+
[source,terminal]
----
How do we kickoff the image gen?
----

. Update the automatically generated `imageBasedUpgrade` CR with the `seedImageRef` field in the target cluster.
+
[source,yaml]
----
apiVersion: lca.openshift.io/v1alpha1
kind: ImageBasedUpgrade
metadata:
  name: upgrade
spec:
  stage: Idle <1>
  seedImageRef: <2>
    version: 4.14.0

>     image: quay.io/<seed_image_repository>

Reviewer 8: ```suggestion
    image: <seed_container_image>
```

// Module included in the following assemblies:
// Epic TELCOSTRAT-160 (4.15/4.16), story TELCODOCS-1576
// * scalability_and_performance/ztp-talm-updating-managed-policies.adoc

:_mod-docs-content-type: PROCEDURE
[id="ztp-image-based-upgrade-seed-generation_{context}"]
= Generating a seed image with the {lcao}

Use the {lcao} to generate the seed image. The image generation does the following:

* Prepares for precaching operations
* Stops the `kubelet` service, running containers, and the CRI-O runtime on the managed cluster
* Backs up platform configurations and artifacts

> * Creates and pushes the OCI images

Reviewer 11: Consider adding a note here that it's expected to lose API availability while LCAO generates the seed image
